src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/CommandProcessor.java

```diff
@@ -2057,4 +2091,33 @@ void executeCommand(Tokens args) {
             }
         }
     }
+
+    /* Parse compression level
+     * @return  0 compress level is out of range
+     *          1 compress level can not be parsed as number
+     */
+    private int parseHeapDumpCompressionLevel(String option) {
+        int gzl = -1;
+        String[] keyValue = option.split("=");
+        if (keyValue[0].equals("gz")) {
+            if (keyValue.length == 1) {
+                err.println("Argument is expected for \"gz\"");
+                return -1;
+            }
+            String level = keyValue[1];
+            try {
+                gzl = Integer.parseInt(level);
+            } catch (NumberFormatException e) {
+                err.println("gz option value not an integer ("+level+")");
+                return -1;
+            }
+            if (gzl < 1 || gzl > 9) {
+                err.println("Compression level out of range (1-9): " + level);
+                gzl = 0;
+            }
+        } else {
+            err.println("Unknow option \"" + option + "\"");
```

Comment:
```
-                if (t.countTokens() > 1) {
+                if (t.countTokens() > 2) {
                     usage();
                 } else {
                     JMap jmap = new JMap();
-                    String filename;
-                    if (t.countTokens() == 1) {
-                        filename = t.nextToken();
-                    } else {
-                        filename = "heap.bin";;
+                    String filename = "heap.bin";
+                    int gzlevel = -1;
+                    int cntTokens = t.countTokens();
```
I'd suggest to move the last line above to the beginning of method and use it in first comparison as well to make it more consistent.

It'd be nice to make the comments more consistent.
For instance:
`+                         * possible command:`
 Start it from capital letter:  `Possible`.
 
 Replace:
```
+                            /* first argument is compression level, second is filename */
+                            /* Parse "gz=" option. */
```
with:
```
+                            /* First argument is compression level, second is filename. */
+                            * Parse "gz=" option. */
```

Replace:
`+                            // Try to parse "gz=" option, if failed, treat it as filename`
with:
`+                            // Try to parse "gz=" option. If failed, treat it as filename.`

Start from capital letter:
`+                                      // compression level not in range.`

```

It'd be nice to reduce the following:
+                    if (cntTokens > 0) {
+                        String option = t.nextToken();
+                        /*
+                         * possible command:
+                         *     dumpheap gz=1 file;
+                         *     dumpheap gz=1;
+                         *     dumpheap file;
+                         */
```

to:
```
+                    String option = t.nextToken();
+                    /*
+                     * Possible command:
+                     *     dumpheap gz=1 file;
+                     *     dumpheap gz=1;
+                     *     dumpheap file;
+                     *     dumpheap;
+                     */
```

It does not seem the condition `if (cntTokens > 0) {`  is very useful.

Also, is it allowed to have option `gz=` without any value referring to the default value?
It seems, Chris already asked a question about max compress level limit.

Reply:
Dear @sspitsyn, 
Thanks for reviewing!  
The following change:
>                              gzlevel = parseHeapDumpCompressionLevel(option)
>                               **if (gzlevel < 1)** {
>                                   err.println("Can not parse compression level from option \"" + option + "\".");
>                                   **if (gzlevel == 0)** {
>                                      // compression level not in range.
>                                       usage();
>                                      return;
>                                  } else {
>                                      out.println("Use \"" + option +"\" as dumped file name.");
>                                   }
>                               } else {
>                                  filename = "heap.bin.gz";
>                               }

The logic here is to handle corner case that use "gz=" as a filename. IMO, dumpheap could work with one argument, and if that argument is not "gz=[number]", it should be treated as filename. 
E.g: "dumpheap gz=", should work to generate a dump file named "gz=", and with some warning message like "Can not parse compression level from option gz=".  
And if the argument is like "gz=10", which is out of range, it will be treated as a wrong gziplevel option and the command will print usage and exit. 

So the code `gzlevel = parseHeapDumpCompressionLeve(option)` is designed to reture 3 type of values:
>  1-9  -- legal comression level
>  0      --  compression level is out of range
> -1     --  other cases, e.g. compression level is not a number. 
Then the above code could process different cases.
P.S.  Found that comments at the definition of parseHeapDumpCompressionLevel() is wrong, maybe it mislead you, sorry for that, and I will fix it at next update.

Thanks,
Lin

Reply:
Hi @sspitsyn,
> It does not seem the condition if (cntTokens > 0) { is very useful.
 
This is used to guard the case cntTokens == 0, otherwise that t.nextToken() will throw ArrayIndexOutOfBoundsException.

Reply:
> This is used to guard the case cntTokens == 0, otherwise that t.nextToken() will throw ArrayIndexOutOfBoundsException.

The line `String option = t.nextToken();` can be moved to each of two if statements:
```
     if (cntTokens == 2) {
         String option = t.nextToken();
         . . .
     } else if (cntTokens == 1) {
         String option = t.nextToken();
         . . .
      }
```
Or you may prefer this:  `String option = (cntTokens > 0) ? t.nextToken() : null;`
Here it is more important to save on statements depth/indentation.

I did not understand your answer on the code below (it seems the problem existed before your fix):
```
                         gzlevel = parseHeapDumpCompressionLevel(option)
                          if (gzlevel < 1) {
                              err.println("Can not parse compression level from option \"" + option + "\".");
                              if (gzlevel == 0) {
                                 // compression level not in range.
                                  usage();
                                 return;
                             } else {
                                 out.println("Use \"" + option +"\" as dumped file name.");
                              }
                          } else {
                             filename = "heap.bin.gz";
                          }
```

There is a check for `gzlevel == 0` when it has been already found this is true: `gzlevel < 1.` It just has no logical sense. I think, the code is broken and has to do something like this:
```
                        } else if (cntTokens == 1) {
                              filename = option;
                             // Try to parse "gz=" option. If failed, treat it as filename.
                             if (option.startsWith("gz=")) {
                                 gzlevel = parseHeapDumpCompressionLevel(option);
                                 if (gzlevel > 0 && gzlevel <= 9) {
                                     // Got correct gzlevel, use default filename.
                                     filename = "heap.bin.gz";
                                 } else {
                                     err.println("Compression level not in range: \"" + option + "\".");
                                     usage();
                                     return;
                                 }
                            }
                        }
```